cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,345 @@
+      Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("notification-publisher-%d")));
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    publishingExecutor.shutdownNow();
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification)
+    throws NotificationException, NotificationFeedException {
+    return publish(feed, notification, notification.getClass());
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(final NotificationFeed feed, final N notification, final Type notificationType)
+    throws NotificationException, NotificationFeedException {
+    LOG.debug("Publishing on notification feed [{}]: {}", feed, notification);
+    return publishingExecutor.submit(new Callable<N>() {
+      @Override
+      public N call() throws Exception {
+        try {
+          KafkaMessage message = new KafkaMessage(KafkaNotificationUtils.buildKafkaMessageKey(feed),
+                                                  GSON.toJsonTree(notification, notificationType));
+          ByteBuffer bb = KafkaMessageCodec.encode(message);
+
+          TopicPartition topicPartition = KafkaNotificationUtils.getKafkaTopicPartition(feed);
+          KafkaPublisher.Preparer preparer = kafkaPublishersCache.get(topicPartition.getTopic());
+          preparer.add(bb, message.getMessageKey());
+
+          try {
+            preparer.send().get();
+            return notification;
+          } catch (InterruptedException e) {
+            throw Throwables.propagate(e);
+          } catch (ExecutionException e) {
+            throw new NotificationException(e.getCause());
+          }
+        } catch (IOException e) {
+          throw new NotificationException(e);
+        }
+      }
+    });
+  }
+
+  private KafkaPublisher getKafkaPublisher() {
+    if (kafkaPublisher != null) {
+      return kafkaPublisher;
+    }
+    try {
+      kafkaPublisher = kafkaClient.getPublisher(ack, Compression.SNAPPY);
+    } catch (IllegalStateException e) {
+      // can happen if there are no kafka brokers because the kafka server is down.
+      kafkaPublisher = null;
+    }
+    return kafkaPublisher;
+  }
+
+
+  @Override
+  public <N> Cancellable subscribe(NotificationFeed feed, NotificationHandler<N> handler)
+    throws NotificationFeedException {
+    return subscribe(feed, handler, null);
+  }
+
+  @Override
+  public <N> Cancellable subscribe(final NotificationFeed feed, final NotificationHandler<N> handler,
+                                   Executor executor) throws NotificationFeedException {
+    // Note: Since we rely on Twill's KafkaConsumer to poll Kafka, and it doesn't
+    // yet allow to use a custom executor, the executor parameter is not used.
+
+    // This call will make sure that the feed exists
+    feedManager.getFeed(feed);
+
+    final TopicPartition topicPartition = KafkaNotificationUtils.getKafkaTopicPartition(feed);
+
+    topicsFeedsLock.writeLock().lock();
```

Comment:
You can actually use the same logic as I showed above for the in memory case. Just treat notification received from Kafka as a publish to an in memory notification service. This can greatly simplify the logic and reduce the usage of lock (especially nested lock).